### PR TITLE
Game save syncing

### PR DESCRIPTION
**Pull request by DominicMaas**
Initial implementation of save save syncing. Still more to do, but that can be done at a later date.

### CODE REVIEW

**Comment by DominicMaas** [comment]
@hibo98, @kaenganxt I've created this draft pull request to help keep track of the implementation of this feature.

**Comment by DominicMaas** [comment]
@hibo98 I didn't know that you had already started (I like our similar ways of implemented this though!), I'll clean up my changes allowing you to merge yours a bit easier (like the sync panel).

Are you on the discord server by any chance?

**Comment by DominicMaas** [comment]
Seems to be working, going to prevent connecting to severs from in game now. Might need to mess around with the order of a few things. e.g the client appears connected before the world is sent (so they are in game and receiving commands, before the actual level is loaded).

**Comment by hibo98** [comment]
@DominicMass I'll have a look, when I'm from EGX Berlin on Monday.

**Comment by lucas-jones** [comment]
https://github.com/DominicMaas/Tango/blob/aa55e9a0c4b5add47e26a80cc2c626211ed6ede1/src/Networking/Client.cs#L258

I've noticed the explanation mark in this string causes the client to crash on connecting to a offline server

**Comment by gregmaxin** [comment]
I would really like to see this implemented. I have some free time over the holidays and would love to put some time into this. 

Whether that be through implementing todos, PR review or play testing... let me know.

**Comment by kaenganxt** [comment]
When hibo's changes are merged, I think we have quite a good implementation for this and we can merge this into the master branch. I think it would be good to have these changes out before christmas (also on Steam) so people can play over the holidays. After that we can take a look to improve this further:
- Allow multiple players to join simultaneously (Requires some logic for showing/hiding the overlay for all players)
- Maybe don't block the whole game while a player is joining (We could only pause the game and queue packets for player interactions on the joining client. I don't think it's possible to let the game run because we don't sync everything perfectly but rely on the simulation threads to be in sync)
- Enable the host to reject join requests
- Allow joining over the Steam friends menu

**Comment by DominicMaas** [comment]
Yeah I think that's a good idea, get this in and working, and aim to optimize it afterwards.

**Comment by DominicMaas** [comment]
Once this is merged, I'll update the steam workshop. I'll aim to merge on Wednesday if no major issues are found.

**Comment by gregmaxin** [comment]
Will this be merged in soon? 

I plan to try playing this mod with my brother again. I may build it myself to test this pr

**Comment by DominicMaas** [comment]
This is a build from this branch that you can use: https://dev.azure.com/dominicmaas/CSM%20-%20Build%20Process/_build/results?buildId=59

And it will be merged fairly soon.

**Comment by gregmaxin** [comment]
Was the goal of this PR to also have periodic game syncing? (i.e every 5min the players download the hosts game save like on join).

I did a playtest and all the changes so far look good, and the game syncing works well.

**Comment by kaenganxt** [comment]
@GodGreg No, periodic syncing is not implemented here. I think this will take some more discussions on how we do things like that. Also thanks for testing this version!

**Comment by kaenganxt** [comment]
My last commit changed the following things:
- Added comments to the commands
- Fixed multiplayer menu not updating on client on first open
- Always go back to main menu on client when disconnecting
- Always run UI update code after server connect in UI thread (otherwise the game could hang)

**Review by kaenganxt — APPROVED** [review]
Tested everything several times, looking good.

**Comment by DominicMaas** [comment]
Awesome to hear, I'll give the latest code changes quick test and we should be good to merge

**Comment by DominicMaas** [comment]
Looks like enabling the mod for the first time does not place the "Join Game" button in the main menu.
If there is not an easy way to fix this, we can always just state that the game MUST be restarted after enabling the mod.

**Comment by kaenganxt** [comment]
I'll see if I can fix it.

**Comment by DominicMaas** [comment]
I'm not sure if this problem was introduced in this branch, but I came across another issue that is located here: #155 (the exception error)

File: src/Panels/JoinGamePanel.cs [hunk 1]
```diff
@@ -71,16 +73,20 @@ public override void Start()
             }
 
             // Add color picker to username field
+            // TODO: Figure out why this is null on main menu
```
Comment (DominicMaas) [inline src/Panels/JoinGamePanel.cs:76]: Keeping a note on this, might be good to see if possible to fix

**Comment by DominicMaas** [comment]
Might also be a good idea to hide the external IP addresses by default, and only show them when clicking/holding/hovering over or something (cases such as steamers for example).

I can take a look into that

**Comment by kaenganxt** [comment]
> Looks like enabling the mod for the first time does not place the "Join Game" button in the main menu.
> If there is not an easy way to fix this, we can always just state that the game MUST be restarted after enabling the mod.

I can only add the button to the bottom of the main menu on mod enable. It may be possible to change the order but I don't see an easy way to do it. Do you think it's okay for now?

> I'm not sure if this problem was introduced in this branch, abut I came across another issue that is located here: #155 (the exception error)

I think I saw that problem also on the master branch, so I don't think it's related to this.

> Keeping a note on this, might be good to see if possible to fix

Since we may still change the player pointers implementation (I think that showing maybe Alt+Mouse clicks to the other players would make more sense, since the current implementation uses quite some network resources), I think that we can leave that for now.

> Might also be a good idea to hide the external IP addresses by default, and only show them when clicking/holding/hovering over or something (cases such as steamers for example).
> 
> I can take a look into that

That's a good idea, but since it's not really related to this feature, can also be done later (Still hoping to get the workshop update out soon :smile:)

**Comment by DominicMaas** [comment]
I think that will be good enough. We can keep the standard code for placing the button in the menu, and have that as an edge case (will we have any cases of two buttons being added?).

I agree on your other points. Once this is merged in, I'll take a look at cleaning up the logging a bit before steam workshop release.

File: src/Panels/JoinGamePanel.cs (hunk 1 above)
Comment (DominicMaas) [reply src/Panels/JoinGamePanel.cs:76]: Taking a look at this later

**Comment by DominicMaas** [comment]
Looks good, all happy for merge?

**Comment by kaenganxt** [comment]
We now use OnEnable/OnDisable to enable/disable our mod as it should be. This handles both cases (game start and enabling through the menu). There are no duplicate buttons as the code first searches for an already existing button.

> Looks good, all happy for merge?

Yeah, it should be fine.

**Pull request merged**